### PR TITLE
Refactoring infra to use bicep registry modules

### PR DESCRIPTION
**Pull request by Gordonby**
## Purpose

To simplify the infrastructure deployment using the bicep registry modules for dapr container apps

Features the following improvements
- Leverages [public bicep modules](https://github.com/Azure/bicep-registry-modules#modules) for simplified infrastructure templating of Dapr Container Apps
- Does not create a dummy app using the nginx image, instead app deployment is delayed until after docker build/push and the ACR has the images
- ~~Resourcetoken is not used, unique strings are generated just for the services that need to be globally unique~~
- Orders and Checkout applications make use of the Infra Output environment variables for a more optimized install

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -8,21 +8,46 @@ Visit [this](https://docs.dapr.io/developing-applications/building-blocks/pubsub
...
+The following prerequisites are required to use this application.  Please ensure that you have them all installed locally.
+
+- [Azure Developer CLI](https://aka.ms/azure-dev/install)
+  - Windows:
+    ```powershell
+    powershell -c "Set-ExecutionPolicy Bypass Process -Force; irm 'https://aka.ms/install-azd.ps1' | iex"
+    ```
+  - Linux/MacOS:
+    ```
+    curl -fsSL https://aka.ms/install-azd.sh | bash
```
Comment (jongio) [inline README.md]: Please update to the latest instructions found on the install page.  Thanks

File: README.md [hunk 2]
```diff
@@ -1,29 +1,54 @@
...
 <!-- END_IGNORE -->
 
+## Quickstart
+
+The fastest way for you to get this application up and running on Azure is to use the `azd up` command. This single command will create and configure all necessary Azure & local resources.
+
+Run the following command to initialize the project, provision the core Azure resources, build the application code in containers and push to a private Azure Container Registry to create a working application.
+
+```bash
+azd up -t Azure-Samples/pubsub-dapr-nodejs-servicebus
```
Comment (jongio) [inline README.md]:
```suggestion
azd up -t pubsub-dapr-nodejs-servicebus
```

File: infra/checkout.bicep [hunk 3]
```diff
@@ -1,104 +1,53 @@
-@minLength(1)
-@maxLength(64)
-@description('Name of the the environment which is used to generate a short unqiue hash used in all resources.')
-param name string
-
-@minLength(1)
-@description('Primary location for all resources')
 param location string
-
+param nameseed string
```
Comment (jongio) [inline infra/checkout.bicep]:
```suggestion
param nameSeed string
```

File: infra/checkout.bicep [hunk 4]
```diff
@@ -1,104 +1,53 @@
...
+  }
+  {
+    name: 'AZURE_KEY_VAULT_ENDPOINT'
+    value: keyvault.properties.vaultUri
+  }
+]
 
-resource keyVault 'Microsoft.KeyVault/vaults@2021-10-01' existing = {
-  name: 'keyvault${resourceToken}'
+resource keyvault 'Microsoft.KeyVault/vaults@2022-07-01' existing = {
```
Comment (jongio) [inline infra/checkout.bicep]:
```suggestion
resource keyVault 'Microsoft.KeyVault/vaults@2022-07-01' existing = {
```

File: infra/containerregistry.bicep [hunk 5]
```diff
@@ -0,0 +1,27 @@
+param location string
+param nameseed string
+param tags object = {}
+
+resource containerRegistry 'Microsoft.ContainerRegistry/registries@2022-02-01-preview' = {
+  name: 'cr${nameseed}${uniqueString(resourceGroup().id, location)}'
```
Comment (jongio) [inline infra/containerregistry.bicep]: Please start using abbreviations.json, which you can find here: https://github.com/Azure/azure-dev/tree/main/templates/common/infra

Example usage here: https://github.com/Azure/azure-dev/blob/main/templates/common/infra/applicationinsights.bicep

Comment (jongio) [inline infra/containerregistry.bicep]: Could you explain further why you have deviated from using resourceToken?  It seems to effectively do the same thing and we'd like to be consistent across samples that are published as canonical samples, so unless there is a REALLY good reason for deviating I would like to remain consistent.

**Review by jongio — CHANGES_REQUESTED** [review]

File: infra/containerregistry.bicep (hunk 5 above)
Comment (Gordonby) [reply infra/containerregistry.bicep]: Looks like they align to CAF naming, so sure. 🤩

Comment (Gordonby) [reply infra/containerregistry.bicep:6]: Resource token use in templates seems to be used to name all resources uniformly to comply with some services requirement to be globally unique.
I don't like it as the primary part of the name, nor its use in services that don't need to be unique.

For the vault and Registry I can pass the resource token to provide the unique element.... but I'm not sure if that would satisfy your point.

File: infra/checkout.bicep (hunk 3 above)
Comment (Gordonby) [reply infra/checkout.bicep]: removed nameseed and replaced with resourceToken

File: infra/containerregistry.bicep (hunk 5 above)
Comment (Gordonby) [reply infra/containerregistry.bicep]: 1. resourceToken does get around the problem of globally unique names, but many Azure resources don't require globally unique names - in these cases it seems to be somewhat of a blunt instrument
1. resourceToken does not result in a meaningful azure resource name for the workload/app. I much prefer the format of [consistent abbreviation][relevant name][optional unique suffix] similar to what you'd see [here](https://docs.microsoft.com/azure/cloud-adoption-framework/ready/azure-best-practices/resource-naming) in the Azure naming best practices.

I've removed my naming approach and fallen back to the resourceToken approach for sake of sample consistency 😄

File: infra/resources.bicep [hunk 6]
```diff
@@ -39,7 +37,7 @@ module containerAppsEnv 'br/public:app/dapr-containerapps-environment:1.0.1' = {
   name: 'caEnv-resources'
```
Comment (jongio) [inline infra/resources.bicep]: caenv-resources

File: infra/checkout.bicep [hunk 7]
```diff
@@ -29,7 +29,7 @@ module appPublisher 'br/public:app/dapr-containerapp:1.0.1' = {
   params: {
     location: location
     containerAppEnvName: containerAppEnvironmentName
-    containerAppName: '${nameseed}checkout'
+    containerAppName: '${name}checkout'
```
Comment (jongio) [inline infra/checkout.bicep]: Just want to confirm that this doesn't need resourceToken?

File: infra/main.bicep [hunk 8]
```diff
@@ -17,6 +17,9 @@ resource resourceGroup 'Microsoft.Resources/resourceGroups@2021-04-01' = {
   location: location
 }
 
+@description('Resource token is used to uniformly name and ensure global name uniqueness for those resources that require it')
+var resourceToken = toLower(uniqueString(subscription().id, name, location))
+
 var tags = {
```
Comment (jongio) [inline infra/main.bicep]: This is a single line in our recent template updates.

File: infra/checkout.bicep [hunk 9]
```diff
@@ -1,104 +1,53 @@
...
+  name: 'dapr-containerapp-publisher'
+  params: {
+    location: location
+    containerAppEnvName: containerAppEnvironmentName
+    containerAppName: '${name}checkout'
+    containerImage: imageName
+    azureContainerRegistry: containerRegistryName
+    environmentVariables: pubSubAppEnvVars
+    enableIngress: false
+    tags: union(tags, {
```
Comment (jongio) [inline infra/checkout.bicep]: Single line...

File: infra/main.bicep [hunk 10]
```diff
@@ -12,41 +12,33 @@ param location string
...
 param principalId string = ''
 
-@description('The image name for the checkout service')
-param checkoutImageName string = ''
-
-@description('The image name for the order-processor service')
-param ordersImageName string = ''
-
 resource resourceGroup 'Microsoft.Resources/resourceGroups@2021-04-01' = {
   name: '${name}-rg'
```
Comment (jongio) [inline infra/main.bicep]: Please use abbrs here and put it as prefix...see our latest in main for ref.

File: infra/main.bicep [hunk 11]
```diff
@@ -12,41 +12,33 @@ param location string
...
+@description('Resource token is used to uniformly name and ensure global name uniqueness for those resources that require it')
 var resourceToken = toLower(uniqueString(subscription().id, name, location))
+
 var tags = {
   'azd-env-name': name
 }
 
 module resources './resources.bicep' = {
-  name: 'resources-${resourceToken}'
+  name: 'resources-${name}'
```
Comment (jongio) [inline infra/main.bicep]: I don't think you need -${name} here...our latest templates don't

File: infra/orders.bicep [hunk 12]
```diff
@@ -1,110 +1,53 @@
...
+  name: 'dapr-containerapp-subscriber'
+  params: {
+    location: location
+    containerAppEnvName: containerAppEnvironmentName
+    containerAppName: '${name}orders'
+    containerImage:  imageName
+    azureContainerRegistry: containerRegistryName
+    environmentVariables: pubSubAppEnvVars
+    targetPort: 5001
+    tags: union(tags, {
```
Comment (jongio) [inline infra/orders.bicep]: single line...

File: infra/checkout.bicep (hunk 7 above)
Comment (Gordonby) [reply infra/checkout.bicep]: I'll try using abbreviations and the resourceToken, but I do recall seeing an azd error when it expected a certain ACA name.

Comment (Gordonby) [reply infra/checkout.bicep]: Hey @jongio 
So this is the azd error i get

> Error: deploying service: deploying service checkout package: fetching service properties: failed running az resource show: exit code: 3, stdout: , stderr: ERROR: (ResourceNotFound) The Resource 'Microsoft.App/containerApps/gordapp6checkout' under resource group 'gordapp6-rg' was not found. For more details please go to https://aka.ms/ARMResourceNotFoundFix
Code: ResourceNotFound
Message: The Resource 'Microsoft.App/containerApps/gordapp6checkout' under resource group 'gordapp6-rg' was not found. For more details please go to https://aka.ms/ARMResourceNotFoundFix
: exit status 3

The resource is created with the resourceToken, and all the Arm deployments succeed.

image

The azd tags are being applied 

image

Looks pretty similar to whats going on here, https://github.com/Azure-Samples/todo-python-mongo-aca/blob/main/infra/api.bicep
So not quite sure why the errors popping out like this.

File: infra/resources.bicep [hunk 13]
```diff
@@ -5,7 +5,7 @@ param tags object
 
 @description('The container registry is used by azd to store your images')
 module registry './containerregistry.bicep' = {
-  name: 'container-registry'
+  name: 'containerRegistry-resources'
```
Comment (jongio) [inline infra/resources.bicep]: we are doing all lower dashes for these names.

File: infra/checkout.bicep (hunk 7 above)
Comment (Gordonby) [reply infra/checkout.bicep]: Reverting to the simpler [Environment][Service] name notation for the Container Apps for the time being.
Have left the abbreviation-service-resourceToken name in as a comment to be addressed in a future PR.

File: infra/resources.bicep [hunk 14]
```diff
@@ -1,98 +1,54 @@
...
-module logAnalyticsResources './loganalytics.bicep' = {
-  name: 'loganalytics-resources'
+@description('The container apps environment is where the applications will be deployed to')
+module containerAppsEnv 'br/public:app/dapr-containerapps-environment:1.0.1' = {
+  name: 'caenv-resources'
   params: {
     location: location
+    nameseed: resourceToken
+    applicationEntityName: 'orders'
+    daprComponentName: 'orderpubsub'
```
Comment (greenie-msft) [inline infra/resources.bicep]: This works fine for the given sample, but what if more than one Dapr component is required?

Comment (Gordonby) [reply infra/resources.bicep]: Then the module in the registry needs to be iterated on to allow that. Is that a use case for this sample?

Comment (greenie-msft) [reply infra/resources.bicep]: No, it's not be we most certainly need to account for that in the module

Comment (Gordonby) [reply infra/resources.bicep]: Cool. 
So, merge time here then?

**Review by greenie-msft — APPROVED** [review]

**Comment by jongio** [comment]
azd core modules have been dramatically updated, you can find latest in main branch.  would be good to determine if this sample should follow that pattern.

**Comment by Gordonby** [comment]
> azd core modules have been dramatically updated, you can find latest in main branch. would be good to determine if this sample should follow that pattern.

I'd prefer to get this one merged and then fast follow to close those gaps with a specific PR.

@greenie-msft is validating if the last commit meets the requirement of abstracting the app naming.

**Comment by Gordonby** [comment]
Closing this down. Due to the time it's taking, the introduction of the azd core modules mean a lot of rework for this template, which is best left for the repo owner.

The infra content of this PR (without azd) can be found in the [Azure Quickstarts](https://github.com/Azure/azure-quickstart-templates/tree/master/quickstarts/microsoft.app/container-app-dapr-pubsub-servicebus)

**Pull request closed without merging**